### PR TITLE
send warning when we have a old commitment transaction

### PR DESCRIPTION
**Pull request by vincenzopalazzo**
During a `channel_reestablish` now we send a warning message when we have an old commitment transaction.

Fixes https://github.com/lightningdevkit/rust-lightning/issues/1207 (but in reality, it is already fixed)

Signed-off-by: Vincenzo Palazzo <vincenzopalazzodev@gmail.com>

### CODE REVIEW

File: lightning/src/ln/channel.rs [hunk 1]
```diff
@@ -3679,6 +3679,21 @@ impl<Signer: Sign> Channel<Signer> {
...
 	}
 
+	/// Check if the we are trying to reestablish a connection when we have a
+	/// very old local commitment transaction.
+	///
+	/// If we are out of sync we return an empty result, otherwise just the channel error tu sent.
+	fn try_reestablish_when_we_are_late(&mut self,  msg: &msgs::ChannelReestablish) -> Result<(), ChannelError> {
+		//FIXME(vincenzopalazzo) We receive the next commitment transaction to one to sent, so to be in sync we should
+		// be equal at list to the `next - 1`?
+		if (INITIAL_COMMITMENT_NUMBER - self.cur_holder_commitment_transaction_number) < msg.next_remote_commitment_number - 1 {
```
Comment (TheBlueMatt) [inline lightning/src/ln/channel.rs:3689]: Heh, you have an integer underflow on this line.

Comment (vincenzopalazzo) [reply lightning/src/ln/channel.rs]: Damnt! Thanks!

Let's see what see the CI now, to check if my actual change fails somewhere (hopefully!)

File: lightning/src/ln/functional_tests.rs [hunk 2]
```diff
@@ -7380,22 +7380,23 @@ fn test_data_loss_protect() {
...
-	check_spends!(node_txn[0], chan.3);
-	assert_eq!(node_txn[0].output.len(), 2);
-	mine_transaction(&nodes[0], &node_txn[0]);
-	connect_blocks(&nodes[0], ANTI_REORG_DELAY - 1);
-	check_closed_event!(nodes[0], 1, ClosureReason::ProcessingError { err: "We have fallen behind - we have received proof that if we broadcast remote is going to claim our funds - we can\'t do any automated broadcasting".to_string() });
-	let spend_txn = check_spendable_outputs!(nodes[0], node_cfgs[0].keys_manager);
-	assert_eq!(spend_txn.len(), 1);
-	check_spends!(spend_txn[0], node_txn[0]);
+	// Clear the event received by the node A, the failure need to be handled by the API user.
+	assert!(!nodes[0].node.get_and_clear_pending_events().is_empty());
```
Comment (vincenzopalazzo) [inline lightning/src/ln/functional_tests.rs:7399]: There are already some why to manage this type of errors? @TheBlueMatt

for now, I just ignored it, because the user API needs to catch why the node send a late.

File: lightning/src/ln/channel.rs [hunk 3]
```diff
@@ -3679,6 +3679,20 @@ impl<Signer: Sign> Channel<Signer> {
 		}
 	}
 
+	/// Check if the we are trying to reestablish a connection with a peer with an old commitment
+	/// transaction that it is not possible recovered.
+	///
+	/// If the peer is not out of sync we return an empty result, otherwise just the channel error to sent.
+	fn try_reestablish_when_peer_is_late(&self, msg: &msgs::ChannelReestablish) -> Result<(), ChannelError> {
```
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply lightning/src/ln/channel.rs:3686]: Well, personally I like to have the logic fragmented into small functions that help to understand what this function does. however, in this case, the logic is too stupid to keep it a function around, maybe in the future if the required code will increase

File: lightning/src/ln/functional_tests.rs (hunk 2 above)
Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs:7399]: I simply take a step mode here. By manage here, I mean that the warning message should be the way to avoid the force close channel when one node is behind.

But this is unclear in the spec, and required more reasoning, maybe in another PR

File: lightning/src/ln/functional_tests.rs [hunk 4]
```diff
@@ -7380,22 +7380,23 @@ fn test_data_loss_protect() {
 	}
 
 	// Check we close channel detecting A is fallen-behind
-	nodes[1].node.handle_channel_reestablish(&nodes[0].node.get_our_node_id(), &reestablish_1[0]);
-	check_closed_event!(nodes[1], 1, ClosureReason::ProcessingError { err: "Peer attempted to reestablish channel with a very old local commitment transaction".to_string() });
-	assert_eq!(check_closed_broadcast!(nodes[1], true).unwrap().data, "Peer attempted to reestablish channel with a very old local commitment transaction");
-	check_added_monitors!(nodes[1], 1);
+	// Check if we sent the warning message when we detecting that A is fallen-behind,
+	// and we give the possibility to A to  be able to recover from error.
+	for msg in nodes[1].node.get_and_clear_pending_msg_events() {
```
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs]: This was fan!

Thanks, I implemented the test case here as you suggested, with the only thing that I use is an enum to difference the test that we want to do.

I found this more easy and readable for this particular case where we can have different types of corner cases, but if you don't like I will use the boolean value.

P.S: I reproduce the ping-point case discussed in https://github.com/lightning/bolts/issues/934#issuecomment-1106752093

File: lightning/src/ln/functional_tests.rs [hunk 5]
```diff
@@ -7282,8 +7282,20 @@ fn test_user_configurable_csv_delay() {
 	} else { assert!(false); }
 }
 
-#[test]
-fn test_data_loss_protect() {
+/// describe the test case as a enum, istead as boolean in this case!
```
Comment (vincenzopalazzo) [inline lightning/src/ln/functional_tests.rs:7285]: I'm too verbose with the comments because I tend to forget what I did. if are too much I will remove it

**Review by vincenzopalazzo — COMMENTED** [review]
OK! I tried to make a little bit of refactoring and applied the changes that you suggested.

The only think that is unclear to me right now, is this case

https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/ln/channel.rs#L3708-L3712

Why in this case there is a `CloseDelayBroadcast` why does mean? i was not able to find some comments about the meaning, and I didn't check if the spec has some info about this case!

File: lightning/src/ln/functional_tests.rs [hunk 6]
```diff
@@ -7380,22 +7392,46 @@ fn test_data_loss_protect() {
...
+	// and we give the possibility to A to  be able to recover from error.
+	expect_warning_due_wrong_channel_reestablish!(nodes[1]);
+
+	// after the waring message sent by B, we should not able to
+	// use the channel, or reconnect with success to the channel.
+	match case {
+		DataLossProtectTestCase::UseChannel => {
+			// try to make a payment, but this should not be allowed
+			// from A -> B
+			let result = fail_to_get_route!(nodes[0], nodes[1], 1000000);
```
Comment (TheBlueMatt) [inline lightning/src/ln/functional_tests.rs:7405]: Instead of a new macro and this verbose check, I think you can just do `assert!(nodes[0].node.list_usable_channels().is_empty())` or so, which is ultimately *why* the route-get fails.

File: lightning/src/ln/functional_test_utils.rs [hunk 7]
```diff
@@ -1203,6 +1225,24 @@ macro_rules! expect_pending_htlcs_forwardable_ignore {
...
 }
 
+#[macro_export]
+macro_rules! expect_warning_due_wrong_channel_reestablish {
+	($node: expr) => {
+		for msg in $node.node.get_and_clear_pending_msg_events() {
+		if let MessageSendEvent::HandleError { ref action, .. } = msg {
+			match action {
+				&ErrorAction::SendWarningMessage { ref msg, .. } => {
+					assert!(msg.data.contains(&"Peer attempted to reestablish channel with a very old local commitment transaction".to_owned()));
```
Comment (TheBlueMatt) [inline lightning/src/ln/functional_test_utils.rs:1235]: Instead of a new macro, why not just extend `check_warn_msg` with the expected message? If not just inline this - I don't think we need to macro-ize this given we're unlikely to need to check this exact error string again.

**Review by TheBlueMatt — COMMENTED** [review]
Sorry for the delay in reviewing this!

File: lightning/src/ln/functional_tests.rs (hunk 6 above)
Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs:7405]: Right, less code is always better!

File: lightning/src/ln/functional_test_utils.rs (hunk 7 above)
Comment (vincenzopalazzo) [reply lightning/src/ln/functional_test_utils.rs:1235]: The only reason that I created a new macro is that I was not able to get working the contains method over an `expr` inside a macro.

The solution can be to simplify the error message to static `str`, where we can call over an `assert_eq!` assert.

What do you think? I'm missing something?

File: lightning/src/ln/functional_test_utils.rs [hunk 8]
```diff
@@ -1190,6 +1190,28 @@ macro_rules! get_route_and_payment_hash {
 	}}
 }
 
+#[cfg(test)]
+#[macro_export]
+macro_rules! fail_to_get_route {
```
Comment (TheBlueMatt) [inline lightning/src/ln/functional_test_utils.rs:1195]: We're deleting this, right? You could also take the `get_route` commit from #1451 as a replacement.

Comment (vincenzopalazzo) [reply lightning/src/ln/functional_test_utils.rs:1195]: >I'm not sure why it didn't work? $expr is generic across any express, as long as all the call sites use the same type it should always work. What did you try and what error did you get?

Yeah, the problem is exactly that `expr` is too generic because contains is only on the string type, strange theory type :/. However, I will commit your suggestion, and in case we revert it

>We're deleting this, right? You could also take the get_route commit from https://github.com/lightningdevkit/rust-lightning/pull/1451 as a replacement.

Right! I had this on my todo list for review, I think yes, having a `get_route` will be helpful, thank you!

File: lightning/src/ln/functional_tests.rs [hunk 9]
```diff
@@ -7380,22 +7392,39 @@ fn test_data_loss_protect() {
...
-	connect_blocks(&nodes[0], ANTI_REORG_DELAY - 1);
-	check_closed_event!(nodes[0], 1, ClosureReason::ProcessingError { err: "We have fallen behind - we have received proof that if we broadcast remote is going to claim our funds - we can\'t do any automated broadcasting".to_string() });
-	let spend_txn = check_spendable_outputs!(nodes[0], node_cfgs[0].keys_manager);
-	assert_eq!(spend_txn.len(), 1);
-	check_spends!(spend_txn[0], node_txn[0]);
+	let mut node_txn = nodes[1].tx_broadcaster.txn_broadcasted.lock().unwrap().clone();
+	// The node B should not broadcast the transaction to force close the channel!
+	assert!(node_txn.is_empty());
+	// B should now detect that there is something wrong and should force close the channel.
+	// FIXME: we sent the warning message when the peer is behind to give the possibility to restore try to fix the problem
```
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs:7418]: Mh! right! this was refused. I moved the FIXME a couple of lines up.

Here my claim is! If we give the possibility to the sender to force close the channel or do what they want in some cases we can fall into some malicious or very buggy code where the sender keeps reconnecting with an old data because he doesn't want to force close the channel.

Why? maybe the code tries to preserve the channel and the spec sais nothing about this case, I would like to see somethings "if the peer keep sending trash the received will force close the channel" somewhere

[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/functional_tests.rs:7418]: A few ways to look at this:

(a) we prefer to use github issues to track things, not in-code FIXMEs or TODOs, so just as a general point would prefer to not add this kind of thing in code
(b) in general the spec shouldn't take a stance on how clients handle cases like this - its totally reasonable for a user to want the option to contact the other node operator out-of-band to avoid an on-chain closure of a big channel or when feerates are high. Its also reasonable for a node to automatically force-close based on a timer to avoid this. Its also reasonable for a node to automatically force-close right away cause the peer is buggy and screw buggy peers. Given there are many potentially correct answers here, the spec should not take a stance.
(c) on the LDK end, we really should make this something the user has more influence over, years ago our stance was to make things as automated as possible but the reality of how people want to use lightning today is somewhat in conflict with that - people get really annoyed when a channel force-closes on them. This is a topic for a separate PR, though.

[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs:7418]: >we prefer to use github issues to track things, not in-code FIXMEs or TODOs, so just as a general point would prefer to not add this kind of thing in code

Oh right, the FIXME was added to point out this in the PR and discuss it. I think this is not really something we should catch in the ldk and also discuss in ldk because the decision of what the node should do is at the user level (from the programmer's viewpoint). So I will remove it when I will push the new fix. I'm working more in deep into the tests to make the tests clean.

>in general the spec shouldn't take a stance on how clients handle cases like this - its totally reasonable for a user to want the option to contact the other node operator out-of-band to avoid an on-chain closure of a big channel or when feerates are high. Its also reasonable for a node to automatically force-close based on a timer to avoid this. Its also reasonable for a node to automatically force-close right away cause the peer is buggy and screw buggy peers. Given there are many potentially correct answers here, the spec should not take a stance.

Right, but in the spec, I would like to see what the node should not do! e.g: Send again the wrong data! BTW this is something to discuss spec discussion, and also something useful to define also to have some global spec integration testing.

**Comment by vincenzopalazzo** [comment]
Rebased on master, I reorganized the test in a more clean way where the different cases are executed at the end of the function. In this way, it is cleaner to the reader what we are trying to do!

File: lightning/src/ln/channel.rs [hunk 10]
```diff
@@ -3715,6 +3715,7 @@ impl<Signer: Sign> Channel<Signer> {
 					if expected_point != PublicKey::from_secret_key(&self.secp_ctx, &given_secret) {
 						return Err(ChannelError::Close("Peer sent a garbage channel_reestablish with secret key not matching the commitment height provided".to_owned()));
 					}
+
```
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply lightning/src/ln/channel.rs:3718]: Sorry!

File: lightning/src/ln/functional_tests.rs [hunk 11]
```diff
@@ -7380,22 +7392,45 @@ fn test_data_loss_protect() {
...
+			assert!(nodes[0].node.list_usable_channels().is_empty());
+		},
+		DataLossProtectTestCase::TryToReconnect => {
+			nodes[0].node.peer_connected(&nodes[1].node.get_our_node_id(), &msgs::Init { features: InitFeatures::empty(), remote_network_address: None });
+			nodes[1].node.peer_connected(&nodes[0].node.get_our_node_id(), &msgs::Init { features: InitFeatures::empty(), remote_network_address: None });
+			let retry_reestablish = get_chan_reestablish_msgs!(nodes[1], nodes[0]);
+
+			nodes[1].node.handle_channel_reestablish(&nodes[0].node.get_our_node_id(), &retry_reestablish[0]);
+			check_added_monitors!(nodes[1], 1);
+			let exp_err = "Peer sent a garbage channel_reestablish with secret key not matching the commitment height provided";
```
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs:7423]: oh boy! I'm stupid.

You have right, I got confused with the test result, and I went stupid around test errors.

In this case, node 0 will not find the channel, and return the error during the `handle_channel_reestablish`, because he was forced to close the channel before.

The question that I have now, is how I can test if node 1 can handle an establishment with a node that has no channel? sound that it is not possible?

Pushing the actual change, maybe help to reason on it

[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs]: Thanks for this tip!

File: lightning/src/ln/functional_tests.rs [hunk 12]
```diff
@@ -7402,22 +7414,63 @@ fn test_data_loss_protect() {
...
 	// Check we close channel detecting A is fallen-behind
+	// Check if we sent the warning message when we detecting that A is fallen-behind,
+	// and we give the possibility to A to  be able to recover from error.
 	nodes[1].node.handle_channel_reestablish(&nodes[0].node.get_our_node_id(), &reestablish_1[0]);
-	check_closed_event!(nodes[1], 1, ClosureReason::ProcessingError { err: "Peer attempted to reestablish channel with a very old local commitment transaction".to_string() });
-	assert_eq!(check_closed_broadcast!(nodes[1], true).unwrap().data, "Peer attempted to reestablish channel with a very old local commitment transaction");
-	check_added_monitors!(nodes[1], 1);
+	let warn_msg = "Peer attempted to reestablish channel with a very old local commitment transaction".to_owned();
+	assert!(check_warn_msg!(nodes[1], nodes[0].node.get_our_node_id(), chan.2).contains(&warn_msg));
+
```
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs:7422]: Sorry!

File: lightning/src/ln/functional_tests.rs [hunk 13]
```diff
@@ -7402,22 +7414,63 @@ fn test_data_loss_protect() {
...
+				}
+			}
+			assert_eq!(err_msgs_0.len(), 1);
+			nodes[1].node.handle_error(&nodes[0].node.get_our_node_id(),&err_msgs_0[0]);
+			assert!(!nodes[1].tx_broadcaster.txn_broadcasted.lock().unwrap().is_empty());
+			assert!(nodes[1].node.list_usable_channels().is_empty());
+			check_added_monitors!(nodes[1], 1);
+			check_closed_event!(nodes[1], 1, ClosureReason::CounterpartyForceClosed { peer_msg: "Failed to find corresponding channel".to_owned() });
+			nodes[1].tx_broadcaster.txn_broadcasted.lock().unwrap().clear();
+			nodes[1].node.get_and_clear_pending_msg_events();
```
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs:7465]: Oh, thanks for this `check_closed_broadcast!(nodes[1], false);` tips

File: lightning/src/ln/functional_tests.rs [hunk 14]
```diff
@@ -7402,22 +7414,61 @@ fn test_data_loss_protect() {
...
+						},
+						_ => panic!("Unexpected event!"),
+					}
+				} else {
+					panic!("Unexpected event!");
+				}
+			}
+			assert_eq!(err_msgs_0.len(), 1);
+			nodes[1].node.handle_error(&nodes[0].node.get_our_node_id(),&err_msgs_0[0]);
+			assert!(!nodes[1].tx_broadcaster.txn_broadcasted.lock().unwrap().is_empty());
```
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs:7459]: ok we can remove the two assert on the `txn_broadcasted`? and do you mean also the check on the usable channels?

[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs:7459]: I got it. Sorry if I ask but I was not sure on what line you were pointing!

[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs]: Done!

File: lightning/src/ln/functional_tests.rs [hunk 15]
```diff
@@ -7402,22 +7414,60 @@ fn test_data_loss_protect() {
...
+	assert!(node_txn.is_empty());
+	// B should now detect that there is something wrong and should force close the channel.
+	let exp_err = "We have fallen behind - we have received proof that if we broadcast remote is going to claim our funds - we can\'t do any automated broadcasting";
+	check_closed_event!(nodes[0], 1, ClosureReason::ProcessingError { err: exp_err.to_string() });
+
+	// after the waring message sent by B, we should not able to
+	// use the channel, or reconnect with success to the channel.
+	match case {
+		DataLossProtectTestCase::UseChannel => {
+			assert!(nodes[0].node.list_usable_channels().is_empty());
```
Comment (valentinewallace) [inline lightning/src/ln/functional_tests.rs:7435]: I'm a bit confused why having these two enum cases isn't identical coverage to moving this line right before the `peer_connected` calls below, and getting rid of the enum?

File: lightning/src/ln/functional_tests.rs [hunk 16]
```diff
@@ -7402,22 +7414,60 @@ fn test_data_loss_protect() {
...
-	assert_eq!(spend_txn.len(), 1);
-	check_spends!(spend_txn[0], node_txn[0]);
+	let mut node_txn = nodes[1].tx_broadcaster.txn_broadcasted.lock().unwrap().clone();
+	// The node B should not broadcast the transaction to force close the channel!
+	assert!(node_txn.is_empty());
+	// B should now detect that there is something wrong and should force close the channel.
+	let exp_err = "We have fallen behind - we have received proof that if we broadcast remote is going to claim our funds - we can\'t do any automated broadcasting";
+	check_closed_event!(nodes[0], 1, ClosureReason::ProcessingError { err: exp_err.to_string() });
+
+	// after the waring message sent by B, we should not able to
```
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs:7431]: Damnt! the grammar checker of my ide start to think like me :)

Sorry about that

File: lightning/src/ln/functional_tests.rs (hunk 15 above)
Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs:7435]: The usage of the enum has a sense when we need to test different corner cases, where avoid using different boolean variables as a function parameter and make the code more readable!

Read this `do_test_data_loss_protect(DataLossProtectTestCase::TryToReconnect);` is better than this `do_test_data_loss_protect(true);` due the missing of name parameter in rust.

However, introducing an enum only for one case is not useful and you have right! I'm removing this. When I introduced it I was expecting that the branching was a little bit more complex!

**Comment by vincenzopalazzo** [comment]
I clean up the code and remove the enum. thanks, @valentinewallace to point me on this, and also for the grammar mistake!

I also rebased on top of the last master!

**Comment by vincenzopalazzo** [comment]
Ops! sorry @TheBlueMatt

File: lightning/src/ln/functional_tests.rs [hunk 17]
```diff
@@ -7402,22 +7401,56 @@ fn test_data_loss_protect() {
...
+	// The node B should not broadcast the transaction to force close the channel!
+	assert!(node_txn.is_empty());
+	// B should now detect that there is something wrong and should force close the channel.
+	let exp_err = "We have fallen behind - we have received proof that if we broadcast remote is going to claim our funds - we can\'t do any automated broadcasting";
+	check_closed_event!(nodes[0], 1, ClosureReason::ProcessingError { err: exp_err.to_string() });
+
+	// after the warning message sent by B, we should not able to
+	// use the channel, or reconnect with success to the channel.
+	assert!(nodes[0].node.list_usable_channels().is_empty());
+	if try_to_reconnect {
```
Comment (valentinewallace) [inline lightning/src/ln/functional_tests.rs:7421]: Sorry I think I was unclear, I think we can get rid of the bool too? Otherwise it seems like the `try_to_reconnect` case doesn't run any unique testing, it'd just be a subset of the `true` case?

Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs:7421]: Oh sorry, I need to move the class also in the `test_*` and remove `do_test_*` function?

Comment (valentinewallace) [reply lightning/src/ln/functional_tests.rs:7421]: Yup exactly my thought

Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs]: Ok, thanks! now should be fine!

File: lightning/src/ln/channel.rs [hunk 18]
```diff
@@ -3737,6 +3737,15 @@ impl<Signer: Sign> Channel<Signer> {
 			}
 		}
 
+		// Before we change the state of the channel, we check if the peer is sending a very old
+		// commitment transaction number, if yes we send a warning message.
+		let our_commitment_transaction = INITIAL_COMMITMENT_NUMBER - self.cur_holder_commitment_transaction_number - 1;
+		if  msg.next_remote_commitment_number + 1 < our_commitment_transaction {
+			return Err(
+				ChannelError::Warn(format!("Peer attempted to reestablish channel with a very old local commitment transaction: {} (received) vs {} (expected)", msg.next_remote_commitment_number, our_commitment_transaction))
```
Comment (valentinewallace) [inline lightning/src/ln/channel.rs:3745]: Oh just realized, I know these are `unreachable`(i think), but for consistency I think it'd be nice to change the other two places we return `Err(ChannelError::Close("Peer attempted to reestablish channel with a very old remote commitment transaction".to_owned()))` to `warnings` as well? @TheBlueMatt

File: lightning/src/ln/functional_tests.rs [hunk 19]
```diff
@@ -7388,8 +7388,7 @@ fn test_data_loss_protect() {
 		if let MessageSendEvent::SendChannelReestablish { ref node_id, ref msg } = msg {
 			assert_eq!(*node_id, nodes[1].node.get_our_node_id());
 			reestablish_1.push(msg.clone());
-		} else if let MessageSendEvent::BroadcastChannelUpdate { .. } = msg {
-		} else if let MessageSendEvent::HandleError { ref action, .. } = msg {
+		} else if let MessageSendEvent::BroadcastChannelUpdate { .. } = msg {} else if let MessageSendEvent::HandleError { ref action, .. } = msg {
```
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply lightning/src/ln/functional_tests.rs:7391]: Sorry, I think that my editor did this edit alone and I didn't check before push!
